connector/rest/src/main/java/it/pagopa/selfcare/dashboard/connector/rest/PartyConnectorImpl.java

```diff
@@ -370,13 +371,24 @@ public void checkExistingRelationshipRoles(String institutionId, String productI
             Set<PartyRole> roles = partyRoleToUsersMap.keySet();
             List<PartyRole> partyRoles = institutionRelationships.stream().map(RelationshipInfo::getRole).collect(Collectors.toList());
 
+            if(checkUserRole(userDto, institutionRelationships)){
+                throw new ValidationException("User role conflict");
+            }
+
             if (!roles.contains(PartyRole.OPERATOR) || !(partyRoles.contains(PartyRole.OPERATOR))) {
                 throw new ValidationException("User role conflict");
             }
         }
         log.trace("checkExistingRelationshipRoles end");
     }
 
+    private boolean checkUserRole(CreateUserDto userDto, RelationshipsResponse institutionRelationships) {
+        List<ProductInfo> productInfo = institutionRelationships.stream().map(RelationshipInfo::getProduct).collect(Collectors.toList());
+        return userDto.getRoles().stream()
+                .anyMatch(userProductRole -> productInfo.stream()
+                        .anyMatch(productInfoRole -> userProductRole.getProductRole().equals(productInfoRole.getRole())));
+    }
+
```

Comment:
current code which can be inefficient if the lists contain a large number of elements. One possible optimization would be to create a set (Set) of product roles from productInfo and then use the contains() method to check if the user's role is present in the set. This approach would have a lower time complexity compared to the current approach.

```
private boolean checkUserRole(CreateUserDto userDto, RelationshipsResponse institutionRelationships) {
    Set<String> productRoles = institutionRelationships.stream()
            .map(RelationshipInfo::getProduct)
            .map(ProductInfo::getRole)
            .collect(Collectors.toSet());

    return userDto.getRoles().stream()
            .map(UserProductRole::getProductRole)
            .anyMatch(productRoles::contains);
}

```